src/Orleans.Runtime/Streams/QueueBalancer/LeaseBasedQueueBalancer.cs

```diff
@@ -113,32 +109,33 @@ public LeaseBasedQueueBalancer(LeaseBasedQueueBalancerOptions options, IServiceP
             this.isStarting = true;
             this.loggerFactory = loggerFactory;
             this.options = options;
-            this.logger = loggerFactory.CreateLogger<LeaseBasedQueueBalancer>();
+            this.logger = loggerFactory.CreateLogger($"{typeof(LeaseBasedQueueBalancer).FullName}-{name}");
         }
 
         public static IStreamQueueBalancer Create(IServiceProvider services, string name, IDeploymentConfiguration deploymentConfiguration)
         {
-            var options = services.GetService<IOptionsSnapshot<LeaseBasedQueueBalancerOptions>>().Get(name);
-            return ActivatorUtilities.CreateInstance<LeaseBasedQueueBalancer>(services, options, deploymentConfiguration);
+            var options = services.GetRequiredService<IOptionsSnapshot<LeaseBasedQueueBalancerOptions>>().Get(name);
+            if (options == null)
+                throw new KeyNotFoundException($"No lease base queue balancer options was configured for provider {name}, nor was a default configured.");
+            return ActivatorUtilities.CreateInstance<LeaseBasedQueueBalancer>(services, name, options, deploymentConfiguration);
         }
         /// <inheritdoc/>
-        public override Task Initialize(string strProviderName, IStreamQueueMapper queueMapper)
+        public override Task Initialize(IStreamQueueMapper queueMapper)
         {
             if (queueMapper == null)
             {
                 throw new ArgumentNullException("queueMapper");
             }
-            var options = this.serviceProvider.GetRequiredService<IOptionsSnapshot<LeaseBasedQueueBalancerOptions>>().Get(strProviderName);
-            if (options == null)
-                throw new KeyNotFoundException($"No lease base queue balancer options was configured for provider {strProviderName}, nor was a default configured.");
-            this.leaseProvider = this.serviceProvider.GetRequiredService(options.LeaseProviderType) as ILeaseProvider;
             this.allQueues = new ReadOnlyCollection<QueueId>(queueMapper.GetAllQueues().ToList());
+            if (this.allQueues.Count == 0)
+                return Task.CompletedTask;
+            this.leaseProvider = this.serviceProvider.GetRequiredService(options.LeaseProviderType) as ILeaseProvider;
             NotifyAfterStart().Ignore();
             //make lease renew frequency to be every half of lease time, to avoid renew failing due to timing issues, race condition or clock difference. 
-            var timerLogger = this.loggerFactory.CreateLogger<AsyncTaskSafeTimer>();
-            this.renewLeaseTimer = new AsyncTaskSafeTimer(timerLogger, this.MaintainAndBalanceQueues, null, this.options.SiloMaturityPeriod, this.options.LeaseLength.Divide(2));
+            ITimerRegistry timerRegistry = this.serviceProvider.GetRequiredService<ITimerRegistry>();
```

Comment:
Is this a good way to create a timer callback that is scheduled on an Orleans thread?

Reply:
Only if you're using it from a grain, but `null` is being passed, which will cause an NRE

Reply:
Oh, I see, TimerRegistry is being modified

Reply:
Yes.  The behavior is that if it's called from an Orleans thread, the callback will be scheduled on an Orleans thread.  If not it won't be.  The grain reference is only necessary for cleanup.  I'm not 100% confident in this change because the api surface doesn't make it clear that a null grain is valid, so other implementations may not account for that.

If we agree that this is a valid behavior (debatable) we may want to modify the interface to make this more clear.

If this behavior is unacceptable, then we need to figure out how to schedule timers on Orleans threads from Orleans threads.

Reply:
Alternatively, it could be argued that it's not the responsibility of the LeaseBasedQueueBalancer to schedule these in Orleans threads, instead making it the responsibility of the lease provider to solve this if it's implementation uses grains.  I solved it this way mainly because this logic _was_ called in an Orleans thread, so making grain calls should have been safe, but was not because of the way the timers were setup.